File: java/sourceafis/src/sourceafis/matching/minutia/HashLookup.java
package sourceafis.matching.minutia;

import java.util.ArrayList;
import java.util.Hashtable;
 
 public final class HashLookup
    {
        EdgeLookup edgeLookup;
        Hashtable<Integer,Object> hash;
        EdgeShape currentCandidate;
        ArrayList<IndexedEdge> currentList;
        int currentOffset;

        public void reset(EdgeHash hash)
        {
            this.edgeLookup = hash.edgeLookup;
            this.hash = hash.hash;
            this.currentList = null;
        }

        @SuppressWarnings("unchecked")
		public IndexedEdge select(EdgeShape candidate)
        {
            //Object value;
            int ihash= edgeLookup.ComputeHash(candidate);
            //if (Hash.TryGetValue(EdgeLookup.ComputeHash(candidate), out value))
            if (hash.containsKey(ihash))
            { 
            	Object value=hash.get(ihash);
                //CurrentList = Hash.get(ihash);//value as List<IndexedEdge>;
                if (value instanceof IndexedEdge)//it can be IndexedEdge or ArrayList
                {
                    IndexedEdge entry = (IndexedEdge)value; 
                    if (edgeLookup.MatchingEdges(entry.shape, candidate))
                        return entry;
                    else
                        return null;
                }
                else// value is ArrayList
                {
                	currentList = (ArrayList<IndexedEdge>)value;
                    for (currentOffset = 0; currentOffset < currentList.size(); ++currentOffset)
                        if (edgeLookup.MatchingEdges(currentList.get(currentOffset).shape, candidate))
                        {
                            currentCandidate = candidate;
                            return currentList.get(currentOffset++);
                        }
                    return null;
                }
            }
            else
                return null;
        }

        public IndexedEdge next()
        {
            if (currentList == null)
                return null;
            while (currentOffset < currentList.size())
            {
                if (edgeLookup.MatchingEdges(currentList.get(currentOffset).shape, currentCandidate))
                    return currentList.get(currentOffset++);
                ++currentOffset;
            }
            return null;
        }
    }

File: SourceAFIS/Matching/HashLookup.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SourceAFIS.Matching
{
    public sealed class HashLookup
    {
        Dictionary<int, object> Hash;
        EdgeShape CurrentCandidate;
        List<IndexedEdge> CurrentList;
        int CurrentOffset;

        public HashLookup(EdgeHash hash)
        {
            Hash = hash.Hash;
        }

        public IndexedEdge Select(EdgeShape candidate)
        {
            object value;
            if (Hash.TryGetValue(EdgeLookup.ComputeHash(candidate), out value))
            {
                CurrentList = value as List<IndexedEdge>;
                if (CurrentList == null)
                {
                    var entry = value as IndexedEdge;
                    if (EdgeLookup.MatchingEdges(entry.Shape, candidate))
                        return entry;
                    else
                        return null;
                }
                else
                {
                    for (CurrentOffset = 0; CurrentOffset < CurrentList.Count; ++CurrentOffset)
                        if (EdgeLookup.MatchingEdges(CurrentList[CurrentOffset].Shape, candidate))
                        {
                            CurrentCandidate = candidate;
                            return CurrentList[CurrentOffset++];
                        }
                    return null;
                }
            }
            else
                return null;
        }

        public IndexedEdge Next()
        {
            if (CurrentList == null)
                return null;
            while (CurrentOffset < CurrentList.Count)
            {
                if (EdgeLookup.MatchingEdges(CurrentList[CurrentOffset].Shape, CurrentCandidate))
                    return CurrentList[CurrentOffset++];
                ++CurrentOffset;
            }
            return null;
        }
    }
}
